src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -345,17 +413,106 @@ public bool Disconnect(DisconnectParams disconnectParams)
             {
-                return false;
-            }
+                // Cancel all pending connections
+                foreach (KeyValuePair<Tuple<string, ConnectionType>, CancellationTokenSource> entry in CancelTupleToCancellationTokenSourceMap)
+                {
+                    string entryConnectionUri = entry.Key.Item1;
+                    ConnectionType entryConnectionType = entry.Key.Item2;
+                    if (disconnectParams.OwnerUri.Equals(entryConnectionUri))
+                    {
+                        CancelConnect(new CancelConnectParams() { OwnerUri = disconnectParams.OwnerUri, Type = entryConnectionType });
+                    }
+                }
 
-            // Lookup the connection owned by the URI
-            ConnectionInfo info;
-            if (!ownerToConnectionMap.TryGetValue(disconnectParams.OwnerUri, out info))
+                // Lookup the connection owned by the URI
+                if (!ownerToConnectionMap.TryGetValue(disconnectParams.OwnerUri, out info))
+                {
+                    return false;
+                }
+
+                // Close all open connections
+                bool closeFailed = false;
+                List<ConnectionType> connectionsToRemove = new List<ConnectionType>();
+
+                foreach (KeyValuePair<ConnectionType, DbConnection> entry in info.ConnectionTypeToConnectionMap)
+                {
+                    try
+                    { 
+                        entry.Value.Close();
+                        connectionsToRemove.Add(entry.Key);
+                    }
+                    catch (System.Data.Common.DbException e)
+                    {
+                        closeFailed = true;
+                    }
+                }
+
+                // If there we no errors calling Close(), simply clear the map and continue
+                if (!closeFailed)
+                {
+                    info.ConnectionTypeToConnectionMap.Clear();
+                    ownerToConnectionMap.Remove(disconnectParams.OwnerUri);
+                }
+                // If there were errors, we will remove only the connections that successfully closed
+                else
+                {
+                    foreach(ConnectionType type in connectionsToRemove)
+                    {
+                        info.ConnectionTypeToConnectionMap.Remove(type);
+                    }
+
+                    return false;
+                }
+
+            }
+            // If we are disconnecting only a single type of connection
+            else
             {
-                return false;
+                // Get non-nullable ConnectionType
+                ConnectionType connectionType = (ConnectionType)disconnectParams.Type;
+
+                // Cancel if we are in the middle of connecting
+                if (CancelConnect(new CancelConnectParams() { OwnerUri = disconnectParams.OwnerUri, Type = connectionType }))
```

Comment:
Separating parts of this out will show that there's a lot of duplication in the "disconnect all" vs "disconnect one" cases. Please simplify this when you're refactoring.